### ISSUE TITLE
[Bug]: a11y addon shows false positives on page load

### ISSUE THREAD

**Issue by jpzwarte** [storybookjs/storybook#30468]
### Describe the bug

When loading the reproduction link, the a11y addon reports 45 problems. If you click the "Tests completed" in the bottom right corner, it will run the tests again and only report 4 problems. This last report is correct (it's about the 4 disabled buttons on the far right).

If you toggle "Highlight results" before you click the "Tests completed", you will also notice that it doesn't clear the highlights of the previous test run.

Reported in #sb-test-early-access channel on Discord.

### Reproduction link

https://kind-plant-088739303-1713.westeurope.1.azurestaticapps.net/?path=/story/actions-button--all&globals=mode:dark

### System

```bash
Storybook Environment Info:

  System:
    OS: macOS 15.3
    CPU: (16) arm64 Apple M4 Max
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.13.1 - /opt/homebrew/bin/node
    Yarn: 4.5.3 - /opt/homebrew/bin/yarn <----- active
    npm: 10.9.2 - /opt/homebrew/bin/npm
  Browsers:
    Chrome: 133.0.6943.54
    Safari: 18.3
  npmPackages:
    @storybook/addon-a11y: ^8.5.3 => 8.5.3 
    @storybook/addon-essentials: ^8.5.3 => 8.5.3 
    @storybook/addon-storysource: ^8.5.3 => 8.5.3 
    @storybook/core-events: ^8.5.3 => 8.5.3 
    @storybook/manager-api: ^8.5.3 => 8.5.3 
    @storybook/preview-api: ^8.5.3 => 8.5.3 
    @storybook/test: ^8.5.3 => 8.5.3 
    @storybook/theming: ^8.5.3 => 8.5.3 
    @storybook/web-components-vite: ^8.5.3 => 8.5.3 
    chromatic: ^11.25.2 => 11.25.2 
    storybook: ^8.5.3 => 8.5.3
```

**Comment by ghengeveld** [comment]
@jpzwarte Can you still reproduce this problem using the [latest 9.0 alpha](https://github.com/storybookjs/storybook/releases)?

Unfortunately your reproduction link is no longer working, so I cannot check the original issue.

> When loading the reproduction link, the a11y addon reports 45 problems. If you click the "Tests completed" in the bottom right corner, it will run the tests again and only report 4 problems. This last report is correct (it's about the 4 disabled buttons on the far right).

This sounds like the A11y report was being generated before the story was completely done rendering/playing. This may happen if there's play function interactions that aren't being `await`-ed, or there's things happening in a setTimeout or something like that. I verified the timing of the report generation, it is being called in the `afterEach` hook which runs after the play function call resolves.

> If you toggle "Highlight results" before you click the "Tests completed", you will also notice that it doesn't clear the highlights of the previous test run.

This should be resolved in https://github.com/storybookjs/storybook/pull/30774 which was released in [v9.0.0-alpha.12](https://github.com/storybookjs/storybook/releases/tag/v9.0.0-alpha.12).

Please let me know if the problem still exists in 9.0 alpha.

**Comment by ghengeveld** [comment]
Based on that example it seems like the A11y report is being generated whilst CSS animations/transitions are still running. The foreground color of the "January 2025" text is changing with every run, and I think that's because the fadein transition it appears to do. Unfortunately it's very hard to predict when CSS transitions will be finished. We can probably rerun it after a second or so to somewhat work around the problem, but more likely we will have to add a `a11y.delay` parameter/global.

**Comment by ghengeveld** [comment]
I talked about this with the team and we think the appropriate solution, at least in the near term, is to use a play function to define your own delay, rather than for us to introduce a delay parameter. It would look something like this:

```ts
play: async () => {
  await new Promise((resolve) => setTimeout(resolve, 1000));
}
```

This would cause the accessibility tests to be delayed by 1 second. The reason is that Storybook waits for the `play` function to finish before running the accessibility tests. In case of dynamically inserted elements and the like, you can use `waitFor` instead:

```
play: async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  await waitFor(async () => {
    await expect(canvas.findByRole('button')).toBeInTheDocument();
  });
}
```

Where `within` and `waitFor` is imported from `@storybook/test` (though it originates from Testing Library).

I would still love to have an "it just works" solution such as an automated delay based on the actual CSS transitions/animations, but that's not a simple task.

**Comment by ghengeveld** [comment]
@jpzwarte @Niznikr This problem should be resolved in v9.0.0 which automatically [waits for animations](https://github.com/storybookjs/storybook/pull/31287). For non CSS animations/transitions, you can use the play function delay shown above.

Please reopen if this doesn't resolve the problem.
